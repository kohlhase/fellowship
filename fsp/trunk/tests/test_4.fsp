lj.

declare A,B : bool.
theorem t1 : (A -> ~A -> B).
 elim a; elim na.
 elim in na b.
 axiom.
qed.

declare C,D : bool.
theorem t2 : (A\/B -> (A -> C)/\(B -> C) -> D -> C/\(D \/ A)).
 elim ab; elim H; elim d.
 elim in ab th.

 elim in H a ac bc (C).
 elim in ac c; axiom.
 elim in th c.
 axiom.
 elim left; axiom.

 elim in H a ac bc (C).
 elim in bc c; axiom.
 elim in th c.
 axiom.
 elim left; axiom.
qed.

checkout coq.
checkout proof term coq.
export natural language.
