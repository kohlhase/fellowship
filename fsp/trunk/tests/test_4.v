(* Coq script *) 
Parameter A B : Prop. 
Theorem t1 : (A -> (~A -> B)). 
Proof (fun (a:A) => (fun (na:~A) => (let H1:~A := na in (match (H1 a) with end)))). 
Parameter C D : Prop. 
Theorem t2 : ((A \/ B) -> (((A -> C) /\  (B -> C)) -> (D -> (C /\  (D \/ A))))). 
Proof (fun (ab:(A \/ B)) => (fun (H:((A -> C) /\  (B -> C))) => (fun (d:D) => (match ab with or_introl x1 => (let a:A := x1 in (match H with (conj ac bc) => (let c:C := (ac a) in (conj c (or_introl _ d))) end)) | or_intror x1 => (let a:B := x1 in (match H with (conj ac bc) => (let c:C := (bc a) in (conj c (or_introl _ d))) end)) end)))). 
