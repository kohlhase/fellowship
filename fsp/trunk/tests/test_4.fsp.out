
    > Current logic: intuitionistic sequent calculus.
      

    > Current logic: intuitionistic sequent calculus.
      
fsp < fsp < 
    > A,B defined.
      
fsp < 
    > Proof term: 
      μthesis:A⇒¬A⇒B.<?1||thesis>
      Natural language:
      we need to prove A⇒¬A⇒B
        .... (1) <======
      done
      
      1 goal yet to prove!
      
      
       |-----  1
      *:A⇒¬A⇒B
      
fsp < 
    > Proof term: 
      μthesis:A⇒¬A⇒B.<λa:A.λna:¬A.?1.1.1||thesis>
      Natural language:
      we need to prove A⇒¬A⇒B
        assume A (a)
        assume ¬A (na)
        .... (1.1.1) <======
      done
      
      1 goal yet to prove!
      
      
      na:¬A
      a:A
       |-----  1.1.1
      *:B
      
fsp < 
    > Proof term: 
      μthesis:A⇒¬A⇒B.<λa:A.λna:¬A.μb:B.<na||μ'H1:¬A.<H1||?1.1.1.2.1*_F_>>||thesis>
      Natural language:
      we need to prove A⇒¬A⇒B
        assume A (a)
        assume ¬A (na)
        we need to prove B
          by na
          we proved ¬A (H1)
          by H1
          and .... (1.1.1.2.1) <======
        absurd
      done
      
      1 goal yet to prove!
      
      
      na:¬A
      a:A
       |-----  1.1.1.2.1
      *:A
      
fsp < 
    > Closed the last branch: 
      Proof completed!
      
      Proof term: 
      μthesis:A⇒¬A⇒B.<λa:A.λna:¬A.μb:B.<na||μ'H1:¬A.<H1||a*_F_>>||thesis>
      Natural language:
      we need to prove A⇒¬A⇒B
        assume A (a)
        assume ¬A (na)
        we need to prove B
          by na
          we proved ¬A (H1)
          by H1
          and by a
        absurd
      done
      
      
fsp < 
    > t1 defined.
      
fsp < fsp < 
    > C,D defined.
      
fsp < 
    > Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<?1||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        .... (1) <======
      done
      
      1 goal yet to prove!
      
      
       |-----  1
      *:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
      
fsp < 
    > Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.?1.1.1.1||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        .... (1.1.1.1) <======
      done
      
      1 goal yet to prove!
      
      
      d:D
      H:(A⇒C)∧(B⇒C)
      ab:A∨B
       |-----  1.1.1.1
      *:C∧(D∨A)
      
fsp < 
    > 2 subgoals generated. 
      
      Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.μth:C∧(D∨A).<ab||(?1.1.1.1.2.1,?1.1.1.1.2.2)>||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        we need to prove C∧(D∨A)
          by ab
          by cases: 
            first case:
              by case hypothesis
            ...(1.1.1.1.2.1) <======
            second case:
              by case hypothesis
            ...(1.1.1.1.2.2)
      done
      
      2 goals yet to prove!
      
      
      *:A
      d:D
      H:(A⇒C)∧(B⇒C)
      ab:A∨B
       |-----  1.1.1.1.2.1
      th:C∧(D∨A)
      
fsp < fsp < 
    > 2 subgoals generated. 
      
      Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.μth:C∧(D∨A).<ab||(μ'a:A.<H||(ac:A⇒C,bc:B⇒C).<?1.1.1.1.2.1.2.1||?1.1.1.1.2.1.2.2>>,?1.1.1.1.2.2)>||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        we need to prove C∧(D∨A)
          by ab
          by cases: 
            first case:
              by case hypothesis
              we proved A (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              .... (1.1.1.1.2.1.2.1) <======
            ...(1.1.1.1.2.1.2.2)
            second case:
              by case hypothesis
            ...(1.1.1.1.2.2)
      done
      
      3 goals yet to prove!
      
      
      ac:A⇒C
      bc:B⇒C
      a:A
      d:D
      H:(A⇒C)∧(B⇒C)
      ab:A∨B
       |-----  1.1.1.1.2.1.2.1
      *:C
      
fsp < 
    > Closed a branch. 
      
      Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.μth:C∧(D∨A).<ab||(μ'a:A.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<ac||a*c>||?1.1.1.1.2.1.2.2>>,?1.1.1.1.2.2)>||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        we need to prove C∧(D∨A)
          by ab
          by cases: 
            first case:
              by case hypothesis
              we proved A (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by ac
                and by a
              done
            ...(1.1.1.1.2.1.2.2) <======
            second case:
              by case hypothesis
            ...(1.1.1.1.2.2)
      done
      
      2 goals yet to prove!
      
      
      *:C
      ac:A⇒C
      bc:B⇒C
      a:A
      d:D
      H:(A⇒C)∧(B⇒C)
      ab:A∨B
       |-----  1.1.1.1.2.1.2.2
      th:C∧(D∨A)
      
fsp < 
    > 2 subgoals generated. 
      
      Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.μth:C∧(D∨A).<ab||(μ'a:A.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<ac||a*c>||μ'c:C.<(?1.1.1.1.2.1.2.2.1.1,?1.1.1.1.2.1.2.2.1.2)||th>>>,?1.1.1.1.2.2)>||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        we need to prove C∧(D∨A)
          by ab
          by cases: 
            first case:
              by case hypothesis
              we proved A (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by ac
                and by a
              done
              we proved C (c)
              .... (1.1.1.1.2.1.2.2.1.1) <======
              and
              .... (1.1.1.1.2.1.2.2.1.2)
            done
            second case:
              by case hypothesis
            ...(1.1.1.1.2.2)
      done
      
      3 goals yet to prove!
      
      
      c:C
      ac:A⇒C
      bc:B⇒C
      a:A
      d:D
      H:(A⇒C)∧(B⇒C)
      ab:A∨B
       |-----  1.1.1.1.2.1.2.2.1.1
      *:C
      
fsp < 
    > Closed a branch. 
      
      Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.μth:C∧(D∨A).<ab||(μ'a:A.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<ac||a*c>||μ'c:C.<(c,?1.1.1.1.2.1.2.2.1.2)||th>>>,?1.1.1.1.2.2)>||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        we need to prove C∧(D∨A)
          by ab
          by cases: 
            first case:
              by case hypothesis
              we proved A (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by ac
                and by a
              done
              we proved C (c)
              by c
              and
              .... (1.1.1.1.2.1.2.2.1.2) <======
            done
            second case:
              by case hypothesis
            ...(1.1.1.1.2.2)
      done
      
      2 goals yet to prove!
      
      
      c:C
      ac:A⇒C
      bc:B⇒C
      a:A
      d:D
      H:(A⇒C)∧(B⇒C)
      ab:A∨B
       |-----  1.1.1.1.2.1.2.2.1.2
      *:D∨A
      
fsp < 
    > Closed a branch. 
      
      Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.μth:C∧(D∨A).<ab||(μ'a:A.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<ac||a*c>||μ'c:C.<(c,L(d))||th>>>,?1.1.1.1.2.2)>||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        we need to prove C∧(D∨A)
          by ab
          by cases: 
            first case:
              by case hypothesis
              we proved A (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by ac
                and by a
              done
              we proved C (c)
              by c
              and
              by d
              trivial
            done
            second case:
              by case hypothesis
            ...(1.1.1.1.2.2) <======
      done
      
      1 goal yet to prove!
      
      
      *:B
      d:D
      H:(A⇒C)∧(B⇒C)
      ab:A∨B
       |-----  1.1.1.1.2.2
      th:C∧(D∨A)
      
fsp < fsp < 
    > 2 subgoals generated. 
      
      Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.μth:C∧(D∨A).<ab||(μ'a:A.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<ac||a*c>||μ'c:C.<(c,L(d))||th>>>,μ'a:B.<H||(ac:A⇒C,bc:B⇒C).<?1.1.1.1.2.2.2.1||?1.1.1.1.2.2.2.2>>)>||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        we need to prove C∧(D∨A)
          by ab
          by cases: 
            first case:
              by case hypothesis
              we proved A (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by ac
                and by a
              done
              we proved C (c)
              by c
              and
              by d
              trivial
            done
            second case:
              by case hypothesis
              we proved B (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              .... (1.1.1.1.2.2.2.1) <======
            ...(1.1.1.1.2.2.2.2)
      done
      
      2 goals yet to prove!
      
      
      ac:A⇒C
      bc:B⇒C
      a:B
      d:D
      H:(A⇒C)∧(B⇒C)
      ab:A∨B
       |-----  1.1.1.1.2.2.2.1
      *:C
      
fsp < 
    > Closed a branch. 
      
      Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.μth:C∧(D∨A).<ab||(μ'a:A.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<ac||a*c>||μ'c:C.<(c,L(d))||th>>>,μ'a:B.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<bc||a*c>||?1.1.1.1.2.2.2.2>>)>||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        we need to prove C∧(D∨A)
          by ab
          by cases: 
            first case:
              by case hypothesis
              we proved A (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by ac
                and by a
              done
              we proved C (c)
              by c
              and
              by d
              trivial
            done
            second case:
              by case hypothesis
              we proved B (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by bc
                and by a
              done
            ...(1.1.1.1.2.2.2.2) <======
      done
      
      1 goal yet to prove!
      
      
      *:C
      ac:A⇒C
      bc:B⇒C
      a:B
      d:D
      H:(A⇒C)∧(B⇒C)
      ab:A∨B
       |-----  1.1.1.1.2.2.2.2
      th:C∧(D∨A)
      
fsp < 
    > 2 subgoals generated. 
      
      Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.μth:C∧(D∨A).<ab||(μ'a:A.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<ac||a*c>||μ'c:C.<(c,L(d))||th>>>,μ'a:B.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<bc||a*c>||μ'c:C.<(?1.1.1.1.2.2.2.2.1.1,?1.1.1.1.2.2.2.2.1.2)||th>>>)>||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        we need to prove C∧(D∨A)
          by ab
          by cases: 
            first case:
              by case hypothesis
              we proved A (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by ac
                and by a
              done
              we proved C (c)
              by c
              and
              by d
              trivial
            done
            second case:
              by case hypothesis
              we proved B (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by bc
                and by a
              done
              we proved C (c)
              .... (1.1.1.1.2.2.2.2.1.1) <======
              and
              .... (1.1.1.1.2.2.2.2.1.2)
            done
      done
      
      2 goals yet to prove!
      
      
      c:C
      ac:A⇒C
      bc:B⇒C
      a:B
      d:D
      H:(A⇒C)∧(B⇒C)
      ab:A∨B
       |-----  1.1.1.1.2.2.2.2.1.1
      *:C
      
fsp < 
    > Closed a branch. 
      
      Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.μth:C∧(D∨A).<ab||(μ'a:A.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<ac||a*c>||μ'c:C.<(c,L(d))||th>>>,μ'a:B.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<bc||a*c>||μ'c:C.<(c,?1.1.1.1.2.2.2.2.1.2)||th>>>)>||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        we need to prove C∧(D∨A)
          by ab
          by cases: 
            first case:
              by case hypothesis
              we proved A (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by ac
                and by a
              done
              we proved C (c)
              by c
              and
              by d
              trivial
            done
            second case:
              by case hypothesis
              we proved B (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by bc
                and by a
              done
              we proved C (c)
              by c
              and
              .... (1.1.1.1.2.2.2.2.1.2) <======
            done
      done
      
      1 goal yet to prove!
      
      
      c:C
      ac:A⇒C
      bc:B⇒C
      a:B
      d:D
      H:(A⇒C)∧(B⇒C)
      ab:A∨B
       |-----  1.1.1.1.2.2.2.2.1.2
      *:D∨A
      
fsp < 
    > Closed the last branch: 
      Proof completed!
      
      Proof term: 
      μthesis:A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A).<λab:A∨B.λH:(A⇒C)∧(B⇒C).λd:D.μth:C∧(D∨A).<ab||(μ'a:A.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<ac||a*c>||μ'c:C.<(c,L(d))||th>>>,μ'a:B.<H||(ac:A⇒C,bc:B⇒C).<μc:C.<bc||a*c>||μ'c:C.<(c,L(d))||th>>>)>||thesis>
      Natural language:
      we need to prove A∨B⇒(A⇒C)∧(B⇒C)⇒D⇒C∧(D∨A)
        assume A∨B (ab)
        assume (A⇒C)∧(B⇒C) (H)
        assume D (d)
        we need to prove C∧(D∨A)
          by ab
          by cases: 
            first case:
              by case hypothesis
              we proved A (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by ac
                and by a
              done
              we proved C (c)
              by c
              and
              by d
              trivial
            done
            second case:
              by case hypothesis
              we proved B (a)
              by H
              we proved A⇒C (ac) and B⇒C (bc)
              we need to prove C
                by bc
                and by a
              done
              we proved C (c)
              by c
              and
              by d
              trivial
            done
      done
      
      
fsp < 
    > t2 defined.
      
fsp < fsp < 
    > Firing up Coq for confirmation.
      
fsp < 
    > Firing up Coq for confirmation.
      
fsp < 
    > File ./test_4.nl written. Enjoy the reading.
      
fsp < 
    > EOF!

